File: harnessOS/sdk/python/harnessos_client/transport.py
"""Transport abstractions for the Python SDK."""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Protocol

from .errors import TransportError


class JsonRpcTransport(Protocol):
    """Synchronous JSON-RPC transport protocol."""

    def request(self, payload: dict[str, Any], *, headers: dict[str, str], timeout: float) -> dict[str, Any]:
        ...


class UrllibJsonRpcTransport:
    """Small stdlib HTTP transport for /v1/rpc."""

    def __init__(self, endpoint: str) -> None:
        self.endpoint = endpoint

    def request(self, payload: dict[str, Any], *, headers: dict[str, str], timeout: float) -> dict[str, Any]:
        body = json.dumps(payload).encode("utf-8")
        request = urllib.request.Request(
            self.endpoint,
            data=body,
            method="POST",
            headers={
                "Content-Type": "application/json",
                **headers,
            },
        )
        try:
            with urllib.request.urlopen(request, timeout=timeout) as response:
                raw = response.read().decode("utf-8")
        except (TimeoutError, urllib.error.URLError, OSError) as exc:
            raise TransportError(f"transport request failed: {type(exc).__name__}") from exc
        try:
            decoded = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise TransportError("transport returned non JSON response") from exc
        if not isinstance(decoded, dict):
            raise TransportError("transport returned malformed JSON-RPC response")
        return decoded

File: harnessOS/sdk/typescript/src/transport.ts
import { TransportError } from "./errors.js";

export interface JsonRpcTransport {
  request(payload: Record<string, unknown>, options: { headers: Record<string, string>; timeoutMs: number }): Promise<unknown>;
}

export class FetchJsonRpcTransport implements JsonRpcTransport {
  readonly endpoint: string;

  constructor(endpoint: string) {
    this.endpoint = endpoint;
  }

  async request(payload: Record<string, unknown>, options: { headers: Record<string, string>; timeoutMs: number }): Promise<unknown> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), options.timeoutMs);
    try {
      const response = await fetch(this.endpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...options.headers,
        },
        body: JSON.stringify(payload),
        signal: controller.signal,
      });
      const text = await response.text();
      try {
        return JSON.parse(text);
      } catch (error) {
        throw new TransportError("transport returned non JSON response");
      }
    } catch (error) {
      if (error instanceof TransportError) throw error;
      throw new TransportError(`transport request failed: ${error instanceof Error ? error.message : String(error)}`);
    } finally {
      clearTimeout(timer);
    }
  }
}
